### PR TITLE
Add functional tests for gunicorn and celery

### PR DESCRIPTION
**Pull request by bloodearnest**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -32,4 +32,8 @@ addons:
   addons:
     postgresql: "9.5"
 install: pip install tox
+before_script:
+ - make db-setup
+services:
+    - redis-server
```
Comment (sparkiegeek) [inline .travis.yml]: inconsistent indentation twitch

File: Makefile [hunk 2]
```diff
@@ -50,13 +50,13 @@ _test: $(VENV)
 
 TEST_FILES = $(shell find tests -maxdepth 1 -name test_\*.py  | cut -c 7- | cut -d. -f1)
 $(TEST_FILES): $(VENV)
-	$(BIN)/pytest -k $@ $(ARGS)
+	. $(BIN)/activate && py.test -k $@ $(ARGS)
```
Comment (sparkiegeek) [inline Makefile]: why this change? shouldn't pytest in the virtualenv already "apply" that virtualenv? maybe it's just the dependency that needs expressing?

File: Makefile [hunk 3]
```diff
@@ -72,6 +72,11 @@ flask: | lib/sqlalchemy
 lib/redis:
 	$(BIN)/pip install redis
 
+db-setup:
+	psql -U postgres -c "create database django_app;"
+	psql -U postgres -c "create user django_app with password 'django_app';"
+	psql -U postgres -c "grant all privileges on database django_app to django_app;"
```
Comment (sparkiegeek) [inline Makefile]: if you create the user first, then you could set the owner of the django_app db to the django_app user and skip the grant?

e.g. CREATE DATABASE django_app OWNER django_app

File: requirements.tests.txt [hunk 4]
```diff
@@ -4,5 +4,9 @@ pytest-cov>=2.3.1
...
 pytest-xdist>=1.22.0
 setuptools>=35.0.0
-eventlet
-gevent
+
+# for integration tests
+eventlet==0.22.0
+gevent==1.2.2
+redis==2.10.6
+SQLAlchemy==1.1.14
```
Comment (sparkiegeek) [inline requirements.tests.txt]: is the strict version selector deliberate? What if downstream projects use different versions?

File: tests/test_functional.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+from talisker.testing import GunicornProcess, ServerProcess
+from tests.celery_app import basic_task, error_task
+
+APP = 'tests.wsgi_app:application'
+
+
+def test_gunicorn_sync_worker():
+    with GunicornProcess(APP, args=['--worker-class=sync']) as p:
+        response = requests.get(p.url('/'))
+        assert response.status_code == 200
```
Comment (sparkiegeek) [inline tests/test_functional.py]: you could move the assert out of the context-manager?

File: tests/test_functional.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
+from builtins import *  # noqa
```
Comment (sparkiegeek) [inline tests/test_functional.py]: hmm? :confused: what's this for?

File: Makefile (hunk 2 above)
Comment (bloodearnest) [reply Makefile]: It's for the venv's $PATH.

The functional tests executes talisker.gunicorn and friends as subprocesses, which if we run the test suite with the venv activated, are on the $PATH. Otherwise we needed some messy path munging to guess the executable path. It's also in line with how tests are run by tox (i.e. in an activated venv), so standardises it.

File: Makefile (hunk 3 above)
Comment (bloodearnest) [reply Makefile]: Huh, TIL, nice

File: requirements.tests.txt (hunk 4 above)
Comment (bloodearnest) [reply requirements.tests.txt]: This is for tests, so downstream won't use it, and is fixed in an upcoming PR anyway

File: tests/test_functional.py (hunk 5 above)
Comment (bloodearnest) [reply tests/test_functional.py]: I could.

File: tests/test_functional.py (hunk 6 above)
Comment (bloodearnest) [reply tests/test_functional.py]: It's to import the future backports in py2. In py3 it does nothing.

https://pypi.python.org/pypi/future

File: requirements.tests.txt (hunk 4 above)
Comment (bloodearnest) [reply requirements.tests.txt]: Fixed as it just specifies a minimum

File: Makefile (hunk 2 above)
Comment (sparkiegeek) [reply Makefile]: :+1: thanks for the explanation

File: tests/test_functional.py (hunk 6 above)
Comment (sparkiegeek) [reply tests/test_functional.py]: Ah, right. Having module names which don't match the package name always confuses me, but it's understandable in this case. Thanks!
